File: targets/widgets/Charts.swift
import SwiftUI

/// One cell of the radial tracker: a wedge of an annulus.
/// Angles are degrees with 0 at 12 o'clock, increasing clockwise — the same
/// convention as RadialTracker.tsx.
struct AnnularSector: Shape {
  let innerRadius: CGFloat
  let outerRadius: CGFloat
  let startAngle: Double
  let endAngle: Double

  func path(in rect: CGRect) -> Path {
    let center = CGPoint(x: rect.midX, y: rect.midY)
    // SwiftUI's 0° is 3 o'clock, so rotate a quarter turn to put 0 at the top
    let start = Angle(degrees: startAngle - 90)
    let end = Angle(degrees: endAngle - 90)

    func point(_ radius: CGFloat, _ angle: Angle) -> CGPoint {
      CGPoint(
        x: center.x + radius * CGFloat(cos(angle.radians)),
        y: center.y + radius * CGFloat(sin(angle.radians))
      )
    }

    var path = Path()
    path.move(to: point(outerRadius, start))
    path.addArc(center: center, radius: outerRadius, startAngle: start, endAngle: end, clockwise: false)
    path.addLine(to: point(innerRadius, end))
    path.addArc(center: center, radius: innerRadius, startAngle: end, endAngle: start, clockwise: true)
    path.closeSubpath()
    return path
  }
}

/// The month-at-a-glance ring chart: one ring per habit, one wedge per day.
struct RadialChart: View {
  let snapshot: PlannerSnapshot
  let theme: Theme
  var minRings: Int = 4
  var showToday: Bool = true

  var body: some View {
    GeometryReader { geo in
      let side = min(geo.size.width, geo.size.height)
      // leave a sliver of room outside the rings for the today marker
      let outer = side / 2 - (showToday ? 5 : 1)
      let inner = side * 0.17
      let rings = max(snapshot.habits.count, minRings)
      let ringWidth = (outer - inner) / CGFloat(rings)
      let days = max(snapshot.daysInMonth, 1)
      let sector = 360.0 / Double(days)
      let gap = min(1.4, sector * 0.09)

      ZStack {
        ForEach(0..<days, id: \.self) { dayIndex in
          ForEach(0..<rings, id: \.self) { ring in
            let r1 = inner + CGFloat(ring) * ringWidth
            let r2 = r1 + ringWidth - max(0.6, ringWidth * 0.12)
            let hasHabit = ring < snapshot.habits.count
            let state = hasHabit ? snapshot.state(day: dayIndex + 1, habitIndex: ring) : 0

            AnnularSector(
              innerRadius: r1,
              outerRadius: r2,
              startAngle: Double(dayIndex) * sector + gap / 2,
              endAngle: Double(dayIndex + 1) * sector - gap / 2
            )
            .fill(theme.stateColor(state))
          }
        }

        if showToday, let today = snapshot.today, today <= days {
          // a dot outside the rings, rather than a wedge across them — a stroked
          // wedge reads as a stray line cutting through the chart
          let mid = (Double(today) - 0.5) * sector - 90
          let rad = mid * .pi / 180
          Circle()
            .fill(theme.accent)
            .frame(width: 4, height: 4)
            .offset(
              x: (outer + 3) * CGFloat(cos(rad)),
              y: (outer + 3) * CGFloat(sin(rad))
            )
        }
      }
      .frame(width: geo.size.width, height: geo.size.height)
    }
    .aspectRatio(1, contentMode: .fit)
  }
}

/// GitHub-style contribution grid: columns are weeks, rows are Sun–Sat.
struct YearGrid: View {
  let snapshot: PlannerSnapshot
  let theme: Theme
  var cell: CGFloat = 7
  var gap: CGFloat = 2
  /// show only the last N weeks (medium widgets can't fit a full year legibly)
  var weekLimit: Int? = nil

  private struct Day {
    let month: Int
    let day: Int
  }

  private var startOffset: Int {
    var comps = DateComponents()
    comps.year = snapshot.year
    comps.month = 1
    comps.day = 1
    let date = Calendar.current.date(from: comps) ?? Date()
    return Calendar.current.component(.weekday, from: date) - 1
  }

  private var days: [Day] {
    var out: [Day] = []
    for m in 0..<12 {
      let len = snapshot.yearDone.indices.contains(m) ? snapshot.yearDone[m].count : 0
      for d in 1...max(len, 1) where len > 0 {
        out.append(Day(month: m, day: d))
      }
    }
    return out
  }

  private func level(_ day: Day) -> Color {
    guard snapshot.yearDone.indices.contains(day.month),
          snapshot.yearDone[day.month].indices.contains(day.day - 1)
    else { return theme.ghLevels[0] }

    let done = snapshot.yearDone[day.month][day.day - 1]
    let missed = snapshot.yearMissed.indices.contains(day.month)
      && snapshot.yearMissed[day.month].indices.contains(day.day - 1)
      ? snapshot.yearMissed[day.month][day.day - 1] : 0
    let habitCount = snapshot.yearHabitCounts.indices.contains(day.month)
      ? snapshot.yearHabitCounts[day.month] : 0

    if done > 0, habitCount > 0 {
      let ratio = Double(done) / Double(habitCount)
      let idx = ratio <= 0.25 ? 1 : ratio <= 0.5 ? 2 : ratio <= 0.75 ? 3 : 4
      return theme.ghLevels[idx]
    }
    if missed > 0 { return theme.missed.opacity(0.45) }
    return theme.ghLevels[0]
  }

  /// Flat day-of-year index of the snapshot's "today", if it has one.
  private var todayIndex: Int? {
    guard let today = snapshot.today else { return nil }
    var idx = 0
    for m in 0..<snapshot.month where snapshot.yearDone.indices.contains(m) {
      idx += snapshot.yearDone[m].count
    }
    return idx + today - 1
  }

  var body: some View {
    let all = days
    let totalCols = Int(ceil(Double(startOffset + all.count) / 7.0))
    // A window has to end at today, not at December — anchoring it to the end
    // of the year would show only empty future weeks for most of the year.
    let lastCol = todayIndex.map { ($0 + startOffset) / 7 } ?? (totalCols - 1)
    let firstCol = weekLimit.map { max(0, lastCol - $0 + 1) } ?? 0
    let endCol = weekLimit == nil ? totalCols : lastCol + 1

    HStack(alignment: .top, spacing: gap) {
      ForEach(firstCol..<endCol, id: \.self) { col in
        VStack(spacing: gap) {
          ForEach(0..<7, id: \.self) { row in
            let idx = col * 7 + row - startOffset
            if idx >= 0, idx < all.count {
              RoundedRectangle(cornerRadius: 1.5)
                .fill(level(all[idx]))
                .frame(width: cell, height: cell)
            } else {
              Color.clear.frame(width: cell, height: cell)
            }
          }
        }
      }
    }
  }
}

/// Progress ring used by the streak and progress widgets.
struct ProgressRing: View {
  let progress: Double
  let theme: Theme
  var lineWidth: CGFloat = 8

  var body: some View {
    ZStack {
      Circle()
        .stroke(theme.cellEmpty, lineWidth: lineWidth)
      Circle()
        .trim(from: 0, to: max(0.001, min(1, progress)))
        .stroke(theme.done, style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
        .rotationEffect(.degrees(-90))
    }
  }
}

/// The section label used across the widgets — small caps with a leading bar.
struct WidgetHeading: View {
  let text: String
  let theme: Theme
  var trailing: String? = nil

  var body: some View {
    HStack(spacing: 5) {
      RoundedRectangle(cornerRadius: 1)
        .fill(theme.accent)
        .frame(width: 3, height: 10)
      Text(text)
        .font(.system(size: 9, weight: .bold))
        .kerning(1.4)
        .foregroundColor(theme.ink)
        .lineLimit(1)
      Spacer(minLength: 2)
      if let trailing {
        Text(trailing)
          .font(.system(size: 9, weight: .semibold))
          .foregroundColor(theme.inkSoft)
          .lineLimit(1)
      }
    }
  }
}

File: targets/android-widgets/kotlin/Charts.kt
package com.owaiskhan.monthlyplanning.widgets

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import java.util.Calendar
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin

/**
 * Charts drawn to bitmaps.
 *
 * Glance has no canvas primitive — it can only lay out a fixed set of views — so
 * anything that isn't a box or a line has to be rasterised here and handed over
 * as an Image. This is the Android counterpart of the SwiftUI `Shape` code in
 * targets/widgets/Charts.swift, and the geometry is deliberately identical.
 */

private fun fillPaint(color: Int) = Paint(Paint.ANTI_ALIAS_FLAG).apply {
  this.color = color
  style = Paint.Style.FILL
}

/**
 * One wedge of an annulus. Android's sweep angles start at 3 o'clock and run
 * clockwise, so every angle is shifted by -90° to put day 1 at the top — the
 * same convention as RadialTracker.tsx.
 */
private fun annularSector(
  cx: Float, cy: Float, r1: Float, r2: Float, startDeg: Float, sweepDeg: Float,
): Path {
  val path = Path()
  val outer = RectF(cx - r2, cy - r2, cx + r2, cy + r2)
  val inner = RectF(cx - r1, cy - r1, cx + r1, cy + r1)
  path.arcTo(outer, startDeg - 90f, sweepDeg)
  path.arcTo(inner, startDeg - 90f + sweepDeg, -sweepDeg)
  path.close()
  return path
}

/** The month as rings: one ring per habit, one wedge per day. */
fun drawRadialChart(
  snapshot: PlannerSnapshot,
  theme: WidgetTheme,
  sizePx: Int,
  minRings: Int = 4,
  showToday: Boolean = true,
): Bitmap {
  val bitmap = Bitmap.createBitmap(sizePx, sizePx, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(bitmap)
  val cx = sizePx / 2f
  val cy = sizePx / 2f
  val markerRoom = if (showToday) sizePx * 0.045f else 2f
  val outer = sizePx / 2f - markerRoom
  val inner = sizePx * 0.17f
  val rings = maxOf(snapshot.habits.size, minRings)
  val ringWidth = (outer - inner) / rings
  val days = maxOf(snapshot.daysInMonth, 1)
  val sector = 360f / days
  val gap = min(1.4f, sector * 0.09f)

  for (dayIndex in 0 until days) {
    val start = dayIndex * sector + gap / 2f
    val sweep = sector - gap
    for (ring in 0 until rings) {
      val r1 = inner + ring * ringWidth
      val r2 = r1 + ringWidth - maxOf(0.6f, ringWidth * 0.12f)
      val state = if (ring < snapshot.habits.size) {
        snapshot.state(dayIndex + 1, ring)
      } else 0
      canvas.drawPath(
        annularSector(cx, cy, r1, r2, start, sweep),
        fillPaint(theme.stateColor(state)),
      )
    }
  }

  // a dot outside the rings rather than a wedge across them
  val today = snapshot.today
  if (showToday && today != null && today <= days) {
    val mid = ((today - 0.5f) * sector - 90f) * Math.PI.toFloat() / 180f
    val radius = outer + markerRoom * 0.6f
    canvas.drawCircle(
      cx + radius * cos(mid),
      cy + radius * sin(mid),
      sizePx * 0.012f,
      fillPaint(theme.accent),
    )
  }
  return bitmap
}

/** GitHub-style grid: columns are weeks, rows are Sun–Sat. */
fun drawYearGrid(
  snapshot: PlannerSnapshot,
  theme: WidgetTheme,
  widthPx: Int,
  cellPx: Float,
  gapPx: Float,
  weekLimit: Int? = null,
): Bitmap {
  val step = cellPx + gapPx
  val height = (7 * step - gapPx).toInt().coerceAtLeast(1)
  val bitmap = Bitmap.createBitmap(widthPx.coerceAtLeast(1), height, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(bitmap)

  val calendar = Calendar.getInstance().apply {
    clear(); set(snapshot.year, 0, 1)
  }
  val startOffset = calendar.get(Calendar.DAY_OF_WEEK) - 1

  data class Day(val month: Int, val day: Int)
  val days = mutableListOf<Day>()
  for (m in 0 until 12) {
    val len = snapshot.yearDone.getOrNull(m)?.size ?: 0
    for (d in 1..len) days.add(Day(m, d))
  }
  if (days.isEmpty()) return bitmap

  val totalCols = ceil((startOffset + days.size) / 7.0).toInt()
  // window has to end at today, not at December, or most of the year it shows
  // nothing but empty future weeks
  val todayIndex = snapshot.today?.let { today ->
    var idx = 0
    for (m in 0 until snapshot.month) idx += snapshot.yearDone.getOrNull(m)?.size ?: 0
    idx + today - 1
  }
  val lastCol = todayIndex?.let { (it + startOffset) / 7 } ?: (totalCols - 1)
  val firstCol = weekLimit?.let { maxOf(0, lastCol - it + 1) } ?: 0
  val endCol = if (weekLimit == null) totalCols else lastCol + 1

  fun levelColor(day: Day): Int {
    val done = snapshot.yearDone.getOrNull(day.month)?.getOrNull(day.day - 1) ?: 0
    val missed = snapshot.yearMissed.getOrNull(day.month)?.getOrNull(day.day - 1) ?: 0
    val habitCount = snapshot.yearHabitCounts.getOrNull(day.month) ?: 0
    if (done > 0 && habitCount > 0) {
      val ratio = done.toDouble() / habitCount
      val idx = when {
        ratio <= 0.25 -> 1
        ratio <= 0.5 -> 2
        ratio <= 0.75 -> 3
        else -> 4
      }
      return theme.ghLevels[idx]
    }
    if (missed > 0) return theme.missed
    return theme.ghLevels[0]
  }

  for (col in firstCol until endCol) {
    for (row in 0 until 7) {
      val idx = col * 7 + row - startOffset
      if (idx < 0 || idx >= days.size) continue
      val left = (col - firstCol) * step
      val top = row * step
      if (left > widthPx) continue
      canvas.drawRoundRect(
        RectF(left, top, left + cellPx, top + cellPx),
        cellPx * 0.22f, cellPx * 0.22f,
        fillPaint(levelColor(days[idx])),
      )
    }
  }
  return bitmap
}

/** Ring showing a 0–1 fraction, matching the SwiftUI ProgressRing. */
fun drawProgressRing(
  fraction: Float,
  theme: WidgetTheme,
  sizePx: Int,
  strokePx: Float,
): Bitmap {
  val bitmap = Bitmap.createBitmap(sizePx, sizePx, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(bitmap)
  val inset = strokePx / 2f
  val rect = RectF(inset, inset, sizePx - inset, sizePx - inset)

  val track = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = theme.cellEmpty
    style = Paint.Style.STROKE
    strokeWidth = strokePx
  }
  canvas.drawArc(rect, 0f, 360f, false, track)

  val progress = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = theme.done
    style = Paint.Style.STROKE
    strokeWidth = strokePx
    strokeCap = Paint.Cap.ROUND
  }
  canvas.drawArc(rect, -90f, 360f * fraction.coerceIn(0.001f, 1f), false, progress)
  return bitmap
}

/**
 * Horizontal progress bar. Drawn rather than composed because Glance has no way
 * to size a child by a fraction of its parent — weights are integers only.
 */
fun drawProgressBar(
  fraction: Float,
  theme: WidgetTheme,
  widthPx: Int,
  heightPx: Int,
): Bitmap {
  val w = widthPx.coerceAtLeast(1)
  val h = heightPx.coerceAtLeast(1)
  val bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(bitmap)
  val radius = h / 2f
  canvas.drawRoundRect(RectF(0f, 0f, w.toFloat(), h.toFloat()), radius, radius, fillPaint(theme.cellEmpty))
  val filled = (w * fraction.coerceIn(0f, 1f)).coerceAtLeast(h.toFloat())
  canvas.drawRoundRect(RectF(0f, 0f, filled, h.toFloat()), radius, radius, fillPaint(theme.accent))
  return bitmap
}

/** Seven-day run of pips used by the streak and quote widgets. */
fun drawPips(
  snapshot: PlannerSnapshot,
  theme: WidgetTheme,
  cellPx: Float,
  gapPx: Float,
): Bitmap? {
  val today = snapshot.today ?: return null
  val first = maxOf(1, today - 6)
  val dayRange = (first..today).toList()
  val width = (dayRange.size * (cellPx + gapPx) - gapPx).toInt().coerceAtLeast(1)
  val bitmap = Bitmap.createBitmap(width, cellPx.toInt().coerceAtLeast(1), Bitmap.Config.ARGB_8888)
  val canvas = Canvas(bitmap)
  dayRange.forEachIndexed { i, day ->
    val left = i * (cellPx + gapPx)
    canvas.drawRoundRect(
      RectF(left, 0f, left + cellPx, cellPx),
      cellPx * 0.25f, cellPx * 0.25f,
      fillPaint(if (snapshot.doneCount(day) > 0) theme.done else theme.cellEmpty),
    )
  }
  return bitmap
}
